course-content/31_exploring_data_structures.qmd

```diff
@@ -404,24 +435,33 @@ We can accomplish this using the `.groupby()` method. A `.groupby()` operation i
 * `prod()`: Product of all items
 * `sum()`: Sum of all items
 
-With our data, let's use these tools to get population mean, min and max by continent. Here's how it works. 
+First let us look at a simple example where we want to get the mean life expectancy for each continent in the data. To do this, we would use the `groupby()` function to call the appropriate segment (i.e. continent), metric (i.e. lifeExp), and type of aggregation (i.e. mean):
 
-First we place "continent" in the `.groupby()` function, then we pass the aggregation functions "mean", "min", and "max" as a dictionary within the `.agg()` function. This dictionary takes the column that we are aggregating - in this case life expectancy - as a key and the aggregation functions as its value.
+```{python}
+simple_mean = country_metrics.groupby('continent').lifeExp.mean()
+simple_mean.head(10)
+```
+
+Fortunately, if we want to look at more aggregations at once, we can do that too. To extend our example above, we would put "continent" in the `.groupby()` function, then pass additional aggregation functions (in our case, we will add  "mean", "min", and "max") as a dictionary within the `.agg()` function. This dictionary takes the column that we are aggregating - in this case life expectancy - as a key and the aggregation functions as its value.
 
 ```{python}
-grouped_single = df.groupby('continent').agg({'lifeExp': ['mean', 'min', 'max']})
+grouped_single = country_metrics.groupby('continent').agg({'lifeExp': ['mean', 'min', 'max']})
 grouped_single.columns = ['lifeExp_mean', 'lifeExp_min', 'lifeExp_max']
-grouped_single.head(10)
+grouped_single.head(5)
 ```
 
 We can nest additional "groups within groups" by creating a list of column names and passing that to the `.groupby()` function instead of passing a single string value.  The example below adds more granularity with the introduction of 'country' and the creation of a list to hold both 'continent' and 'country'.  
 
 ```{python}
-grouped_multiple = df.groupby(['continent', 'country']).agg({'pop': ['mean', 'min', 'max']})
+grouped_multiple = country_metrics.groupby(['continent', 'country']).agg({'pop': ['mean', 'min', 'max']})
 grouped_multiple.columns = ['pop_mean', 'pop_min', 'pop_max']
```

Comment:
This line needs an explanation I think -> otherwise they will forget to do it and end up with multi-indexed rows and columns which is way more confusing to work with (same way that `.reset_index()` needs an explanation)

Reply:
I did more detail on reset_index() above in the section where i looked at sorting

Reply:
The line where you change the column names is what needs explaining as well.

Reply:
Good call, added some explanation for the column naming

Reply:
@swayneh do you mean the addition on line 527? `We also add a line of code that gives each of the columns a new name with the `.columns()` function.`